sources/packages/api/src/services/student-assessment/student-assessment.service.ts

```diff
@@ -399,9 +405,57 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
         "status",
       )
       .innerJoin("assessment.application", "application")
+      .leftJoin("assessment.studentAppeal", "studentAppeal")
+      .leftJoin(
+        "assessment.studentScholasticStanding",
+        "studentScholasticStanding",
+      )
       .where("application.id = :applicationId", { applicationId })
       .orderBy("status", "DESC")
       .addOrderBy("assessment.submittedDate", "DESC")
-      .getRawMany();
+      .getRawAndEntities();
+
+    return mapFromRawAndEntities<AssessmentHistory>(queryResult, "status");
+  }
+
+  /**
+   * Checks if some student assessment is still being processed.
+   * Only one student assessment can be processed at a given time because
+   * any assessment can generate Over Awards and they must be taken into
+   * the consideration every time.
+   * * Alongside with the check, the DB has an index to prevent that a new
+   * * assessment record is created when there is already one with the
+   * * assessment data not populated (submitted/pending).
+   * @param userId user id to check the assessments.
+   * @returns true if there is an assessment that is not finalized yet.
+   */
+  async hasIncompleteAssessment(userId: number): Promise<boolean> {
+    const queryResult = await this.repo
+      .createQueryBuilder("assessment")
+      .select("1")
+      .innerJoin("assessment.application", "application")
+      .innerJoin("application.student", "student")
+      .innerJoin("student.user", "user")
+      .where("user.id = :userId", { userId })
+      .andWhere("assessment.assessmentData IS NULL")
+      .limit(1)
+      .getRawOne();
+    return !!queryResult;
+  }
+
+  /**
+   * Validate if the student has any student assessment that it is not
+   * finished yet (submitted/pending). If there is a student assessment
+   * already being processed, throws an exception.
+   * @param userId user id to check the assessments.
+   */
+  async assertAllAssessmentsCompleted(userId: number) {
```

Comment:
Why did we create the query in different method to check the logic.? is hasIncompleteAssessment been used anywhere other than assertAllAssessmentsCompleted?

Reply:
It will be used also by @ann-aot in her upcoming PR and anywhere else that needs to check before starting a new assessment. For now, only for _Student Appeal_ during its approval and for _Scholastic Standing_ upon its creation.

Reply:
Ah. Got it. Any such thing in future can have a TODO comment to help the reviewers understand and the developer who is going to leverage the method can delete the TODO while doing their story.